CorsixTH/Lua/endconditions.lua

```diff
@@ -0,0 +1,323 @@
+-- or if findbest is true, the score of the group.
+function EndConditions:checkWinGroup(hospital, table, findbest)
+  local met_count, total_count, gap
+  for crit_name, crit_table in pairs(table) do
+    met_count, total_count = 0, 0
+    gap = WinCriteria[crit_name](hospital, crit_table)
+    if gap then
+      met_count = met_count + 1
+    end
+    total_count = total_count + 1
+  end
+
+  if findbest then
+    if met_count > 0 and total_count > 0 then
+      return met_count / total_count
+    else return 0
+    end
+  -- Have all criteria of the group have been met?
+  elseif met_count == total_count then
+    return "win"
+  end
+end
+
+-- Find the group of win conditions best met by the hospital.
+--!param hospital (Hospital) The hospital of the tests.
+--!return best (number) The number of the best group.
+function EndConditions:findBestWinGroup(hospital)
+  local score, best = 0, 1
+  for group = 1, self.highest_group do
+    if self.win_goals[group] then
+      local test = self:checkWinGroup(hospital, self.win_goals[group], true)
+      if test and test > score then best = group end
+    end
+  end
+  -- Return the group number of the group that meets its total most
+  return best
+end
+
+
+class "LoseCriteria"
+---@type LoseCriteria
+--!param hospital (Hospital) The hospital of the tests.
+--!param table (table) the lose_criteria table for one group and one criteria.
+--!return If report is false, return the gap from hospital value to lose value, and the lose value
+-- if true, return the gap from the hospital value to boundary, and the boundary.
+function LoseCriteria.reputation(hospital, table, report)
+  local target = report and table.boundary or table.lose_value
+  local rep = hospital.reputation
+  local max_min = table.max_min_lose == 1 and 1 or -1
+  if (rep - target) * max_min >= 0 then
+    return math.abs(rep - target), target
+  end
+end
+
+function LoseCriteria.balance(hospital, table, report)
+  local target = report and table.boundary or table.lose_value
+  local bal = hospital.balance - hospital.loan
+  local max_min = table.max_min_lose == 1 and 1 or -1
+  if (bal - target) * max_min >= 0 then
+    return math.abs(bal - target), target
+  end
+end
+
+function LoseCriteria.percentage_killed(hospital, table, report)
+  local target = report and table.boundary or table.lose_value
+  local percentage_killed = hospital.percentage_killed
+  local max_min = table.max_min_lose == 1 and 1 or -1
+  if (percentage_killed - target) * max_min >= 0 then
+    return math.abs(percentage_killed - target), target
+  end
+end
+
+
+class "WinCriteria"
+--@type WinCriteria
+
+--!param hospital (Hospital) The hospital of the tests.
+--!param table (table) the win_criteria table for one group and one criteria
+--!return If report is false, return the gap from hospital value to win value, and the win value.
+-- if true, return the gap from the hospital value to boundary, and the boundary.
```

Comment:
On doc, this only will apply to the top function. I fully understand why you made this decision, but for our documentation system it leaves gaps for sure.

Reply:
If you merge the function computations into one as suggested elsewhere, this doc can be attached to that function.